status/status.go

```diff
@@ -58,6 +58,21 @@ func (se *statusError) GRPCStatus() *Status {
 	return &Status{s: (*spb.Status)(se)}
 }
 
+// Is implements future error.Is functionality.
+// A statusError is equivalent if the code and message are identical.
+func (se *statusError) Is(target error) bool {
+	if target == nil {
+		return false
+	}
+
+	tse, ok := target.(*statusError)
+	if !ok {
+		return false
+	}
+
+	return se.Code == tse.Code && se.Message == tse.Message
```

Comment:
This doesn't check the `Details` field -- to do so, I believe you can use `proto.Equal(se, tse)` instead of comparing the fields individually.

Reply:
@dfawley I believe we don't actually want to check the `Details` field, as we very well might have different details added onto the same error "type"

I imagine `WithDetails` could be used for things like tagging something like the userId, or host source, which would break the equivalency.

Reply:
The same types of arguments could be used for checking the Message field as well.  It's common for users to put debugging details (username, raw request information, etc) into RPC error strings.  Whether this is more or less likely is conjecture.  I believe the spirit of "Is" is to perform an exact match, not a fuzzy one.  But I'll have to look at the errors proposal again.

Reply:
Thinking about this some more, I think since Details is just appended to the end of the message, equality would break if they were added in a different order, which doesn't seem great.

Reply:
I'm still of the opinion that it should be matching on just the code and message, but I'm happy to do a strict comparison instead if that will help push this through.

Reply:
From the [errors proposal](https://go.googlesource.com/proposal/+/master/design/29934-error-values.md):

> The Is function follows the chain of errors by calling Unwrap, searching for one that matches a target. **It is intended to be used instead of equality for matching sentinel errors (unique error values)**

If the goal is to match sentinel errors, then details should be the same.  Note that details are not necessarily extraneous debugging information, they may describe the type of failure, e.g. the [example for `PreconditionViolation` in the googleapi's `error_details.proto`](https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto#L90): terms of service not accepted.

It isn't usually correct to use `Is` to match RPC statuses in general, because it isn't common to use sentinal errors in RPC handlers; typically users should manually check the important parts of the status instead (e.g. only the code, or look for a particular `Details` entry like the `RetryInfo`).

IMO even more important functionality from the new errors proposal would be `As`.  We don't export `statusError`, so `As` cannot be used directly, like:

```go
se := &status.statusError{}
ok := errors.As(err, se)
```

We could implement `As` on `*statusError` so that it will work as follows:

```go
if s, ok := &status.Status{}, errors.As(err, s); ok {
	// use s
}
```

And/or provide a convenience wrapper:

```go
if s, ok := status.Unwrap(err); ok {
	// use s; Unwrap implemented with errors.As(err, *statusError)
	// Function name up for debate.  Or possibly implement this in FromError().
}
```

What do you think?

Reply:
Your `Is` argument makes sense to me, thanks for the explanation!

For `As`, I'm unsure about implementing it on &status.Status, as this isn't actually an error type, and it might be an anti-pattern to have `As` work on a non-error type? Maybe they intend for it to be usable this way but it seems a little weird.

Alternatively, we could export a `interface{ GRPCStatus() *status.Status }` type to be used in `As`.

For the convenience wrapper, it seems like it would make sense to implement it in `FromError()`, although that seems like it would have to be in a separate go2 package. I'm not sure if they have a plan for helping packages transition to go2; maybe they'll provide a built-in build directive?

Reply:
> For As, I'm unsure about implementing it on &status.Status, as this isn't actually an error type, and it might be an anti-pattern to have As work on a non-error type? Maybe they intend for it to be usable this way but it seems a little weird.

I agree this seems a bit weird.  `As` would need to be implemented on `*statusError`, but it would assign to a `*status.Status`.  I'm not sure if this would be considered an anti-pattern, but I will pose the question.

> Alternatively, we could export a interface{ GRPCStatus() *status.Status } type to be used in As.

I considered this as well.  I'm not sure this option would be intuitive or discoverable, and I'd rather just offer a `status.Unwrap(error)` helper instead.

> For the convenience wrapper, it seems like it would make sense to implement it in FromError(), although that seems like it would have to be in a separate go2 package.

We would have to move `FromError` into its own file, and define two versions, one for pre-1.13 versions of Go and one for 1.13+.  It might be undesirable/unexpected to unwrap in that function automatically, although I'm having trouble seeing where it would cause problems to do so.

Reply:
> I'm not sure if this would be considered an anti-pattern, but I will pose the question.

Did you get a chance to think about this more?

Your plan for `FromError` seems good.

Reply:
Sorry, I have not had a chance to ask yet.  I'll see if I can find something out soon.

Reply:
`errors.Is()` compares every single element in a chain of errors by `Unwrap()`ing the next level.
`WithDetails()` just adds such an element to the chain, so why should it be part of the comparison?

It just uses a different way for adding an element than the `error` package, but the principle is the same.

Reply:
That's an interesting thought, but the problem is that `WithDetails` isn't necessarily wrapping a different error - it can also explain more about the error itself.  We can't simply ignore it in a function intended to test strict equality.